p2p/net/swarm/dial_ranker.go

```diff
@@ -77,70 +75,92 @@ func DefaultDialRanker(addrs []ma.Multiaddr) []network.AddrDelay {
 	for i := 0; i < len(addrs); i++ {
 		res = append(res, network.AddrDelay{Addr: addrs[i], Delay: 0})
 	}
+
 	res = append(res, getAddrDelay(pvt, PrivateTCPDelay, PrivateQUICDelay, 0)...)
-	res = append(res, getAddrDelay(ip4, PublicTCPDelay, PublicQUICDelay, 0)...)
-	res = append(res, getAddrDelay(ip6, PublicTCPDelay, PublicQUICDelay, 0)...)
+	res = append(res, getAddrDelay(public, PublicTCPDelay, PublicQUICDelay, 0)...)
 	res = append(res, getAddrDelay(relay, PublicTCPDelay, PublicQUICDelay, relayOffset)...)
 	return res
 }
 
-// getAddrDelay ranks a group of addresses(private, ip4, ip6) according to the ranking logic
-// explained in defaultDialRanker.
+// getAddrDelay ranks a group of addresses according to the ranking logic explained in
+// documentation for defaultDialRanker.
 // offset is used to delay all addresses by a fixed duration. This is useful for delaying all relay
-// addresses relative to direct addresses
+// addresses relative to direct addresses.
 func getAddrDelay(addrs []ma.Multiaddr, tcpDelay time.Duration, quicDelay time.Duration,
 	offset time.Duration) []network.AddrDelay {
+
 	sort.Slice(addrs, func(i, j int) bool { return score(addrs[i]) < score(addrs[j]) })
 
+	// If the first address is (QUIC, IPv6), make the second address (QUIC, IPv4).
+	happyEyeballs := false
+	if len(addrs) > 0 {
+		if isQUICAddr(addrs[0]) && isProtocolAddr(addrs[0], ma.P_IP6) {
+			for i := 1; i < len(addrs); i++ {
+				if isQUICAddr(addrs[i]) && isProtocolAddr(addrs[i], ma.P_IP4) {
+					addrs[i], addrs[1] = addrs[1], addrs[i]
```

Comment:
Is swapping the right thing to do here, or should the address be inserted at position 1, shifting all other addresses?

Reply:
There's no impact because we have three tiers: 
IPv6 dialed first, IPv4 dialed second, rest of the addresses. 
I'll change it to shifting though because that seems more correct in case we eventually want to dial only a single address at a time.

Reply:
updated